spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java

```diff
@@ -0,0 +1,275 @@
+        .filter(kv -> kv.getValue().size() > 1)
+        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    // Nothing to rewrite if there's only one DataFile in each partition.
+    if (filteredGroupedTasks.isEmpty()) {
+      return RewriteDataFilesActionResult.empty();
+    }
+
+    long openFileCost = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.SPLIT_OPEN_FILE_COST,
+        TableProperties.SPLIT_OPEN_FILE_COST_DEFAULT);
+
+    // Split and combine tasks under each partition
+    List<CombinedScanTask> combinedScanTasks = filteredGroupedTasks.values().stream()
+        .map(scanTasks -> {
+          CloseableIterable<FileScanTask> splitTasks = TableScanUtil.splitFiles(
+              CloseableIterable.withNoopClose(scanTasks), targetSizeInBytes);
+          return TableScanUtil.planTasks(splitTasks, targetSizeInBytes, splitLookback, openFileCost);
+        })
+        .flatMap(Streams::stream)
+        .collect(Collectors.toList());
+
+    JavaRDD<CombinedScanTask> taskRDD = sparkContext.parallelize(combinedScanTasks, combinedScanTasks.size());
+
+    Broadcast<FileIO> io = sparkContext.broadcast(fileIO);
+    Broadcast<EncryptionManager> encryption = sparkContext.broadcast(encryptionManager);
+
+    RowDataRewriter rowDataRewriter =
+        new RowDataRewriter(table, spec, caseSensitive, io, encryption, targetSizeInBytes);
+
+    List<DataFile> addedDataFiles = rowDataRewriter.rewriteDataForTasks(taskRDD);
+    List<DataFile> currentDataFiles = filteredGroupedTasks.values().stream()
+        .flatMap(tasks -> tasks.stream().map(FileScanTask::file))
+        .collect(Collectors.toList());
+    replaceDataFiles(currentDataFiles, addedDataFiles);
+
+    return new RewriteDataFilesActionResult(currentDataFiles, addedDataFiles);
+  }
+
+  private Map<StructLikeWrapper, Collection<FileScanTask>> groupTasksByPartition(
+      CloseableIterator<FileScanTask> tasksIter) {
+    ListMultimap<StructLikeWrapper, FileScanTask> tasksGroupedByPartition = Multimaps.newListMultimap(
+        Maps.newHashMap(), Lists::newArrayList);
+
+    try {
+      tasksIter.forEachRemaining(task -> {
+        StructLikeWrapper structLike = StructLikeWrapper.wrap(task.file().partition());
+        tasksGroupedByPartition.put(structLike, task);
+      });
+
+    } finally {
+      try {
+        tasksIter.close();
+      } catch (IOException ioe) {
+        LOG.warn("Failed to close task iterator", ioe);
+      }
+    }
+
+    return tasksGroupedByPartition.asMap();
+  }
+
+  private void replaceDataFiles(Iterable<DataFile> deletedDataFiles, Iterable<DataFile> addedDataFiles) {
+    boolean threw = true;
+
+    try {
+      RewriteFiles rewriteFiles = table.newRewrite();
+      rewriteFiles.rewriteFiles(Sets.newHashSet(deletedDataFiles), Sets.newHashSet(addedDataFiles));
+      commit(rewriteFiles);
+      threw = false;
+
+    } finally {
+      if (threw) {
```

Comment:
Can we simplify this block by using `catch (Exception e)` and rethrowing the original exception? Otherwise, we won't know what happened.

Reply:
We typically use this pattern with a `threw` boolean to always clean up even if the throwable isn't an `Exception`. It's more important to do this when the finally block needs to execute for all cases, though. I think it would be probably be fine to catch `Exception` here like it was originally -- sorry for the bad review comment.